asserts/database.go

```diff
@@ -24,23 +24,52 @@ package asserts
 import (
 	"errors"
 	"fmt"
-	"io/ioutil"
-	"net/url"
 	"os"
 	"path/filepath"
 	"regexp"
 	"strings"
 	"time"
-
-	"github.com/ubuntu-core/snappy/helpers"
 )
 
+// BuilderFromComps can build an assertion from its components.
+type BuilderFromComps func(headers map[string]string, body, content, signature []byte) (Assertion, error)
+
+// Backstore is a backstore for assertions. It can store and retrieve
+// assertions by type under primary paths (tuples of strings). Plus it
+// supports more general searches.
+type Backstore interface {
+	// Init initializes the backstore. It is provided with a function
+	// to build assertions from their components.
+	Init(buildAssert BuilderFromComps) error
+	// Put stores an assertion under the given unique primaryPath.
+	// It is responsible for checking that assert is newer than a
+	// previously stored revision.
+	Put(assertType AssertionType, primaryPath []string, assert Assertion) error
```

Comment:
primaryPath should probably be primaryKeys, to give more legroom for the backstore to implement indexing and the such on top of the assertion data without having to hold the concept of a path externally to it.